File: gold_prediction_linear_regression/prediction_model.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define MAX_DATA_POINTS 1000

void normalize(double *data, int n, double *min_val, double *max_val) {
    *min_val = data[0];
    *max_val = data[0];
    for (int i = 1; i < n; i++) {
        if (data[i] < *min_val) *min_val = data[i];
        if (data[i] > *max_val) *max_val = data[i];
    }
    for (int i = 0; i < n; i++) {
        data[i] = (data[i] - *min_val) / (*max_val - *min_val);
    }
}

int main() {
    double x[MAX_DATA_POINTS], y[MAX_DATA_POINTS];
    int n = 0;
    FILE *file = fopen("data.csv", "r");
    if (file == NULL) {
        printf("Error: Unable to open file.\n");
        return 1;
    }

    // Load data from CSV
    while (fscanf(file, "%lf,%lf", &x[n], &y[n]) == 2) {
        n++;
        if (n >= MAX_DATA_POINTS) break;
    }
    fclose(file);

    if (n == 0) {
        printf("Error: Dataset is empty.\n");
        return 1;
    }

    // Normalize data
    double min_x, max_x, min_y, max_y;
    normalize(x, n, &min_x, &max_x);
    normalize(y, n, &min_y, &max_y);

    // Gradient Descent
    double m = 0.0, b = 0.0;
    double learning_rate = 0.01;
    int iterations = 1000;

    for (int i = 0; i < iterations; i++) {
        double dm = 0.0, db = 0.0;
        for (int j = 0; j < n; j++) {
            double prediction = m * x[j] + b;
            double error = prediction - y[j];
            dm += (2.0 / n) * error * x[j];
            db += (2.0 / n) * error;
        }

        if (isnan(dm) || isnan(db)) {
            printf("Numerical instability detected at iteration %d. dm: %f, db: %f\n", i, dm, db);
            break;
        }

        m -= learning_rate * dm;
        b -= learning_rate * db;
    }

    // Denormalize coefficients
    m = m * (max_y - min_y) / (max_x - min_x);
    b = b * (max_y - min_y) + min_y;

    printf("Trained Model: y = %fx + %f\n", m, b);

    // Predict
    double input;
    printf("Enter a value to predict: ");
    if (scanf("%lf", &input) != 1) {
        printf("Invalid input.\n");
        return 1;
    }

    double normalized_input = (input - min_x) / (max_x - min_x);
    double prediction = m * normalized_input + b;
    printf("Predicted value: %f\n", prediction);

    return 0;
}

File: gold_prediction_linear_regression/prediction_model.py
import pandas as pd
import numpy as np

def read_dataset(filename):
    """
    Reads a CSV file with two columns and returns two numpy arrays: x (independent variable) and y (dependent variable).
    """
    try:
        data = pd.read_csv(filename, header=None, names=['x', 'y'])
        x = data['x'].values
        y = data['y'].values
        return x, y
    except Exception as e:
        print(f"Error reading the dataset: {e}")
        return None, None

def normalize(data):
    """
    Normalizes the data to the range [0, 1].
    """
    return (data - np.min(data)) / (np.max(data) - np.min(data))

def linear_regression(x, y, learning_rate=0.01, iterations=1000):
    """
    Performs linear regression using gradient descent.

    Args:
        x: Independent variable.
        y: Dependent variable.
        learning_rate: Step size for gradient descent.
        iterations: Number of iterations to run gradient descent.

    Returns:
        m: Slope of the line.
        b: Intercept of the line.
    """
    m = 0.0  # Initial slope
    b = 0.0  # Initial intercept
    n = len(x)

    for i in range(iterations):
        # Predictions and errors
        predictions = m * x + b
        errors = predictions - y

        # Gradients
        dm = (2 / n) * np.dot(errors, x)
        db = (2 / n) * np.sum(errors)

        # Update parameters
        m -= learning_rate * dm
        b -= learning_rate * db

        # Compute loss and print for debugging
        loss = np.mean(errors ** 2)
        if i % 100 == 0:  # Print every 100 iterations
            print(f"Iteration {i}: Loss = {loss:.4f}, m = {m:.4f}, b = {b:.4f}")

        # Stop if gradients become invalid
        if np.isnan(dm) or np.isnan(db):
            print("NaN detected in gradients. Stopping gradient descent.")
            break

    return m, b

def predict(x, m, b):
    """
    Predicts the value of y given x, slope (m), and intercept (b).

    Args:
        x: Input value.
        m: Slope.
        b: Intercept.

    Returns:
        Predicted value of y.
    """
    return m * x + b

if __name__ == "__main__":
    # Read dataset
    filename = "data.csv"
    x, y = read_dataset(filename)
    if x is None or y is None:
        print("Failed to load dataset.")
        exit(1)

    # Normalize data
    x = normalize(x)
    y = normalize(y)

    # Train model
    learning_rate = 0.01
    iterations = 1000
    m, b = linear_regression(x, y, learning_rate, iterations)

    print(f"Trained Model: y = {m:.2f}x + {b:.2f}")

    # Predict for a new input
    try:
        user_input = float(input("Enter a value to predict: "))
        normalized_input = (user_input - np.min(x)) / (np.max(x) - np.min(x))
        prediction = predict(normalized_input, m, b)
        denormalized_prediction = prediction * (np.max(y) - np.min(y)) + np.min(y)
        print(f"Predicted value: {denormalized_prediction:.2f}")
    except ValueError:
        print("Invalid input. Please enter a numeric value.")
